### PR TITLE
Export improvements

### LINKED ISSUE #657: Improve export CLI
### Kind of request

Changing existing functionality

### Enhancement Description

I had a quick look at what was done in PR #627. Nice start!
I have a few suggestions or things to discuss before we also officially add this function to our docs in #644

First a bug:

- region argument is not passed to the export_data function (and should be converted to bbox first)

Then comments:

- check typos in the docstrings 
- in the docstrings, I feel using '.' for the mandatory path argument instead of path/to/output_dir is less clear. Also compared to other command line examples for build/update
- CLI args option name: -f is used for config file while in build/update we use -i. Do you think we should harmonize and use -i here as well?
- CLI args option name: here might be completely my taste but would -s instead of -t for source names be possible? -s sounds maybe closer to data source compared to target?
- add CLI args option: would it be possible to add an option in CLI for time tuple? Then in really most cases, you don't need a config file, unless for advanced options.
- support all arguments of the export data functions from the config file (unit_add, meta, append)
- nice-to-have: support multiple sources option (current -t) from CLI like we support multiple data catalogs option (-d) in build/update
- format of the config file: might be taste again but I feel that separating args and sources does not feel so clear to me.. could maybe all arguments be put on the same level and naming following a little closer the arguments of the export_data function? Also not sure why dictionary is necessary for each new source? (also data_libs is used instead of catalog in build/update and I am not sure what is the tuple syntax in yaml). My idea would be something like:

```
export:
  region: 
    - subbasin: [-7.24, 62.09]
    - uparea: 50
  data_libs:
    - tests/data/test_sources.yml
  time_tuple: (2010-01-01, 2020-12-31)
  source_names:
    - hydro_lakes[precip]
    - gtsmv3_eu_era5
  unit_conversion: False
  append: False
  meta: 
    version: 0.1
```

### Use case

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by savente93**
## Issue addressed
Fixes #657 

## Explanation
- I ended up not allowing multiple instances of time tuples because that would end up either overwriting the previous exports, or changing the file names, which I didn't want to do. 
- Other changes are based on the discussion Hélène and I had. 

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `main`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed
- [x] Updated changelog.rst if needed

### CODE REVIEW

File: docs/changelog.rst [hunk 1]
```diff
@@ -13,13 +13,16 @@ Unreleased
 
 Added
 -----
+- Export CLI now also accetps time tuples (#660)
```
Comment (hboisgon) [inline docs/changelog.rst:16]:
```suggestion
- Export CLI now also accepts time tuples (#660)
```

File: docs/user_guide/data_overview.rst [hunk 2]
```diff
@@ -61,6 +61,23 @@ A special exception is made for the Deltares data catalog which can be accessed
     data sources with the same names, the sources from the last catalog in the list are used.
     If the ``--dd (--deltares-data)`` flag is used the deltares_data catalog is read first.
 
+The CLI also provides two other utility subcommands: `export` and `check`. THe export function can be used
```
Comment (hboisgon) [inline docs/user_guide/data_overview.rst:64]:
```suggestion
The CLI also provides two other utility subcommands: `export` and `check`. The export function can be used
```

File: docs/user_guide/data_overview.rst [hunk 3]
```diff
@@ -61,6 +61,23 @@ A special exception is made for the Deltares data catalog which can be accessed
...
+to export a subset of the data avaialble in your data catalogs for e.g. collaboration. This sub command supports
```
Comment (hboisgon) [inline docs/user_guide/data_overview.rst:65]:
```suggestion
to export a subset of the data available in your data catalogs for e.g. collaboration. This sub command supports
```

File: docs/user_guide/data_overview.rst [hunk 4]
```diff
@@ -61,6 +61,23 @@ A special exception is made for the Deltares data catalog which can be accessed
...
+providing a region (bounding box or path to geom only), data catalogs, data catalgo entry names, and time ranges
```
Comment (hboisgon) [inline docs/user_guide/data_overview.rst:66]:
```suggestion
providing a region (bounding box or path to geom only), data catalogs, data source names, and time ranges
```

File: docs/user_guide/data_overview.rst [hunk 5]
```diff
@@ -61,6 +61,23 @@ A special exception is made for the Deltares data catalog which can be accessed
...
+for example:
+
+.. code-block:: console
+
+    hydromt export -r "{'bbox': [4.6891,52.9750,4.9576,53.1994]}" -s era5_hourly -d ../hydromt/data/catalogs/artifact_data.yml -t '["2010-01-01", "2022-12-31"]' path/to/output_dir
+
+
+Finally the CLI also provides a lightweight command to check your files for compliance without having to build the model.
+Note that this does not yet validate the data specified, but only the syntax of the files themselfes.
```
Comment (hboisgon) [inline docs/user_guide/data_overview.rst:75]:
```suggestion
Note that this does not yet validate the data specified, but only the syntax of the files themselves.
```

File: docs/user_guide/data_overview.rst [hunk 6]
```diff
@@ -61,6 +61,23 @@ A special exception is made for the Deltares data catalog which can be accessed
...
+However, we do plan to expand it's functionality over time. It can be use for example like so:
```
Comment (hboisgon) [inline docs/user_guide/data_overview.rst:76]:
```suggestion
However, we do plan to expand its functionality over time. It can be use for example like so:
```

File: hydromt/cli/main.py [hunk 7]
```diff
@@ -440,14 +440,15 @@ def export(
     Example usage:
     --------------
 
-    export the data of in a single source, in a pertcular region
-    hydromt export -r "{'subbasin': [-7.24, 62.09], 'uparea': 50}" -t era5_hourly -d ../hydromt/data/catalogs/artifact_data.yml .
+    export the data of in a single source, in a particular region, in a particular time range
```
Comment (hboisgon) [inline hydromt/cli/main.py:443]:
```suggestion
    export the data of a single data source, in a particular region, for a particular time range
```

File: hydromt/cli/main.py [hunk 8]
```diff
@@ -440,14 +440,15 @@ def export(
...
+    hydromt export -r "{'bbox': [4.6891,52.9750,4.9576,53.1994]}" -s era5_hourly -d ../hydromt/data/catalogs/artifact_data.yml -t '["2010-01-01", "2022-12-31"]' path/to/output_dir
+    note to sam: only support BBOX and GEOM
```
Comment (hboisgon) [inline hydromt/cli/main.py:445]:
```suggestion

```

File: hydromt/cli/main.py [hunk 9]
```diff
@@ -440,14 +440,15 @@ def export(
...
 
-    export all data of in a single source
-    hydromt export --dd -t era5_hourly .
+    export all data of in a single source of the deltares data catalog
```
Comment (hboisgon) [inline hydromt/cli/main.py:447]:
```suggestion
    export a single data source from the deltares data catalog without time/space slicing
```

File: hydromt/cli/main.py [hunk 10]
```diff
@@ -440,14 +440,15 @@ def export(
...
+    hydromt export --dd -s era5_hourly path/to/output_dir
```
Comment (hboisgon) [inline hydromt/cli/main.py:448]: I think we want to deprecate the --dd flag:
```suggestion
    hydromt export -d deltares_data -s era5_hourly path/to/output_dir
```

File: hydromt/cli/main.py [hunk 11]
```diff
@@ -464,31 +465,58 @@ def export(
...
+
+    if source:
+        if isinstance(source, str):
+            sources = [source]
+        else:
+            sources = list(source)
+
+    if config:
+        config_dict = cli_utils.parse_config(config)["export_data"]
+        args = config_dict.pop("args", {})
```
Comment (hboisgon) [inline hydromt/cli/main.py:478]: Is 'args' still in the export_data config? If not need to update below

File: hydromt/cli/main.py [hunk 12]
```diff
@@ -464,31 +465,58 @@ def export(
...
         if "catalog" in args.keys():
```
Comment (hboisgon) [inline hydromt/cli/main.py:479]: I suggested data_libs in the issue instead of catalog in the template (not as a bullet point). I leave up to you to change but data_libs is used in the build and update command to specify catalogs from config file ([global] section of the config corresponding to Model.__init__ keys)

File: hydromt/cli/main.py [hunk 13]
```diff
@@ -464,31 +465,58 @@ def export(
...
 
     try:
-        data_catalog = DataCatalog(data_libs=data_libs)
         data_catalog.export_data(
             export_dest_path,
-            source_names=export_targets,
-            time_tuple=time_tuple,
+            source_names=sources,
+            bbox=bbox,
+            time_tuple=time_tup,
```
Comment (hboisgon) [inline hydromt/cli/main.py:519]: I miss passing the other arguments from config_dict like unit_conversion etc. (especially if you pop the others, can easily added here as **config_dict)

**Review by hboisgon — CHANGES_REQUESTED** [review]
Almost there @savente93, thanks for looking into it! 
I think the multiple -t options is actually implemented (it was support multiple sources which at the time were using -t instead of -s now :) )
What I miss is passing the extra arguments from config and for you to check if the config is properly parsed by your function still with the changes implemented

**Comment by savente93** [comment]
> What I miss is passing the extra arguments from config and for you to check if the config is properly parsed by your function still with the changes implemented

Sorry I don't understand what you are asking here, could you rephrase it for me? other than that, all of the comments should be resolved now.

File: hydromt/cli/main.py (hunk 13 above)
Comment (hboisgon) [reply hydromt/cli/main.py:531]: Oh sorry, what you did also works perfectly :) I meant pass here export_data(source_names = sources, ..., **config_dict) and instead of defining unit_conversion = True etc. just define config_dict = dict() in the case there is no config file provided.
But his also works perfectly well
